src/main/java/com/conveyal/datatools/common/utils/Scheduler.java

```diff
@@ -1,63 +1,203 @@
+        for (Project project : Persistence.projects.getAll()) {
+            if (project.autoFetchFeeds) {
+                scheduleAutoFeedFetch(project, 1);
+            }
+        }
+        LOG.info("Scheduling feed expiration notifications");
+        // Get all active feed sources
+        for (FeedSource feedSource : Persistence.feedSources.getAll()) {
+            // Schedule expiration notification jobs for the latest feed version
+            scheduleExpirationNotifications(feedSource);
+        }
+    }
+
+    /**
+     * Convenience method for scheduling one-off jobs.
+     */
+    public static ScheduledJob scheduleJob (String id, MonitorableJob job, long delay, TimeUnit timeUnit) {
+        ScheduledFuture scheduledFuture = schedulerService.schedule(job, delay, timeUnit);
+        ScheduledJob scheduledJob = new ScheduledJob(job, scheduledFuture);
+        scheduledJobsById.put(id, scheduledJob);
+        return scheduledJob;
+    }
 
-        scheduleFeedNotificationExpirations();
+    /**
+     * Convenience method for scheduling recurring jobs.
+     */
+    public static <T extends MonitorableJob> ScheduledJob scheduleRecurringJob (String id, T job, long delay, long period, TimeUnit timeUnit) {
+        ScheduledFuture scheduledFuture = schedulerService.scheduleAtFixedRate(job, delay, period, timeUnit);
+        ScheduledJob scheduledJob = new ScheduledJob(job, scheduledFuture);
+        scheduledJobsById.put(id, scheduledJob);
+        return scheduledJob;
     }
 
     /**
-     * Schedule recurring auto-fetches for all projects as needed
+     * Cancels and removes all scheduled jobs for a given id and job class.
      */
-    private static void scheduleProjectAutoFetches() {
-        for (Project project : Persistence.projects.getAll()) {
-            if (project.autoFetchFeeds) {
-                ScheduledFuture scheduledFuture = ProjectController.scheduleAutoFeedFetch(project, 1);
-                autoFetchMap.put(project.id, scheduledFuture);
+    public static void removeJobsOfType(String id, Class<?> clazz, boolean mayInterruptIfRunning) {
+        int jobsCancelled = 0;
+        // First get the list of jobs belonging to the id (e.g., all jobs related to a feed source).
+        List<ScheduledJob> scheduledJobs = scheduledJobsById.get(id);
+        // Iterate over jobs, cancelling and removing only those matching the job class.
+        for (ScheduledJob scheduledJob : scheduledJobs) {
+            if (clazz.isInstance(scheduledJob.job)) {
+                scheduledJob.scheduledFuture.cancel(mayInterruptIfRunning);
+                scheduledJobsById.remove(id, scheduledJob);
+                jobsCancelled++;
             }
         }
+        // De-clutter the logs, but only logging when jobs have actually been cancelled/removed
+        if (jobsCancelled > 0) {
+            LOG.info("Cancelled/removed {} {} jobs for {}", clazz.getSimpleName(), jobsCancelled, id);
```

Comment:
The clazz and jobsCancelled should probably be switched here.  Also, logging the id is going to be kinda ambiguous since it may not be clear what the id is of.  Is it a Project, FeedSource, something else?

Reply:
cc @abyrd. This is related to https://github.com/catalogueglobal/datatools-server/pull/130#discussion_r238890048